### ISSUE TITLE
IWYU wants to remove includes for types that are needed in template default values

### ISSUE THREAD

**Issue by honkstar1** [include-what-you-use/include-what-you-use#1147]
Here's an example:

```
#include "BaseClass.h"

template<typename TheSizeType>
struct TSizeLargerThan4
{
	static inline constexpr bool TheValue = sizeof(TheSizeType) > 4;
};

template <typename TheType, bool bIsSizeLarger = TSizeLargerThan4<TheType>::TheValue>
struct TGetCastType2
{
	static inline constexpr bool SizeValue = bIsSizeLarger;
};

void myfunc()
{
	auto v = TGetCastType2<BaseClass>::SizeValue;
	(void)v;
}

```

In the above code IWYU wants to turn BaseClass into a forward declaration (but it is needed for the sizeof())

Trying to figure out where this code lives but again, happy for pointers.

**Comment by honkstar1** [comment]
seems like the SubstNonTypeTemplateParmExpr is not implemented at all? I know nothing about llvm and clang implementation but I added a visitor for this and log it out and I see the sizeof(TheType) in there.. trying to figure out how the non-logging path is handling but could need some help :-)

**Comment by kimgr** [comment]
Templates are pretty tricky in IWYU, partly, I think, because its AST visitor predates the `shouldVisitTemplateInstantiations` in Clang. I don't have a clear idea either exactly how it works :).

But my high-level understanding is that when IWYU finds a template instantiation, it starts a new AST traversal (`InstantiatedTemplateASTVisitor`), which visits all types in the instantiation and tries to map them back to the actually-written ("sugared") types provided as template arguments (I'm probably slaughtering this completely).

AFAIK, the handling of default template arguments is in the `GetTplTypeResugarMapForClass` family of functions, so you might want to start there.

**Comment by honkstar1** [comment]
I think I kind of know where the problem is.. but don't know how to solve it.
The InstantiatedTemplateASTVisitor need to also go in to the default expression of the NonTypeTemplateParmDecl found in SubstNonTypeTemplateParmExpr.

So I've added TraverseSubstNonTypeTemplateParmExpr and get the call.. trying to figure out how to have InstantiatedTemplateASTVisitor get past the stmt and in to the UnaryExprOrTypeTraitExpr that is underneath..

**Comment by kimgr** [comment]
I can see that that would make sense as a working theory. I can't say whether it's the right one :)

The `Traverse` methods are used to modify traversal order/granularity whereas `Visit...` is called for every node visited in traversal. So I wonder if you should/could override `VisitUnaryExprOrTypeTraitExpr` in instantiated-template-visitor, and act directly? Probably not, you'd probably need to backtrack and see if you're part of a `SubstNonType...` in that case. If you haven't already done so, try bumping verbosity level to 6 or 7 when running IWYU to get an idea of how it sees the AST.

**Comment by honkstar1** [comment]
haha, so I got it working but I have no idea if it is the correct solution.

I had to do three things.. here's the ones
1. Fix so resugar_map_ contains the non-specialized template type to specialized type. (tbh, I thought that was the purpose of it?.. right now it contains SameType -> SameType). With my code it nows also contains TheType -> BaseClass. 
2. Added to InstantiatedTemplateVisitor::TraverseUnaryExprOrTypeTraitExpr to try to Resugar the type, and if it succeeds, report type use.
3. Add handling for SubstNonTypeTemplateParmExpr by adding InstantiatedTemplateVisitor::TraverseSubstNonTypeTemplateParmExpr and in there also visit the default argument.

Here's the different snippets
1. Look inside GetTplTypeResugarMapForClassNoComponentTypes. The for loop is the new part
```
  const NamedDecl* decl = TypeToDeclAsWritten(tpl_spec_type);
  const auto* cls_tpl_decl = dyn_cast<ClassTemplateSpecializationDecl>(decl);
  if (cls_tpl_decl) {
	  
    const TemplateArgumentList& tpl_arg_list =
        cls_tpl_decl->getTemplateInstantiationArgs();
    tpl_args = tpl_arg_list.asArray();
    num_args = tpl_arg_list.size();

	int index = 0;
	for (NamedDecl* param_decl : *cls_tpl_decl->getSpecializedTemplate()->getTemplateParameters())
	{
		if (auto templ_param_decl = dyn_cast<clang::TemplateTypeParmDecl>(param_decl))
		{
			auto template_type = templ_param_decl->getTypeForDecl();
			retval[template_type] = GetTemplateArgAsType(tpl_args[index]);
		}
		++index;
	}
  }
```

2. Look in TraverseUnaryExprOrTypeTraitExpr. The "else if" statement is the new part
```
else if (const  Type* resugared_type = ResugarType(arg_type))
{
      ReportTypeUse(caller_loc(), resugared_type);
}
```

3. TraverseSubstNonTypeTemplateParmExpr added to InstantiatedTemplateVisitor
```
  bool TraverseSubstNonTypeTemplateParmExpr(SubstNonTypeTemplateParmExpr* arg) {
	const clang::NonTypeTemplateParmDecl* decl = arg->getParameter();
	if (decl->hasDefaultArgument())
      if (!TraverseStmt(decl->getDefaultArgument()))
        return false;

	return Base::TraverseSubstNonTypeTemplateParmExpr(arg);
  }
```

What do you think? Lol, it feels correct but tbh. Since the specialization inside clang turns the sizeof(Type) to a value the specialization doesnt even contain that template anymore. As far as I can tell there is no other way but to keep track of conversion from templated type to instantiated templated type. And then re-run on the non-type-template parameter expressions.

**Comment by honkstar1** [comment]
Unfortunately this solution did not work with recursion. The inner templated type does not get its template arguments in the resugar list.. so it can't lookup the "real" type. 

Maybe this is the wrong approach. It feels correct and that there should be a TemplateSpecialization entry in the AST.

Or maybe it takes care of the first template first and make a record that it replays in some way? I am not sure how it works :-/

(yes I'm using high verbosity)

**Comment by honkstar1** [comment]
aaah, now i understand how it works. One InstantiatedTemplateVisitor is created and then the resugar table is overwritten for every new template type.

So if we have

```
template <typename IndirectedType, unsigned long TheSize = sizeof(IndirectedType)>
struct Foo
{
	static inline constexpr unsigned long Value = TheSize;
};

template <typename TheType>
struct TGetCastType2
{
	static inline constexpr bool SizeValue = Foo<TheType>::Value;
};

```
Then it will first specialize Foo on "TheType" which is not a real type yet. It will then specialize TGetCastType2 later when it is used and traverse the ast assuming it can translate IndirectedType to TheType to whatever TheType is specialized on. But since the resugar logic is overwritten entirely when specializing TGetCastType it will lose the translation from IndirectedType to TheType.... so when we enter the sizeof(IndirectedType) we have lost the translation from Indirected type.

How well is IWYU handling nested templates etc? Either I have misunderstood something fundamental or I can imagine lots of scenarios that break. It feels like the former :-)

**Comment by honkstar1** [comment]
tested this too.. didn't work. It wants to remove BaseClass.

```
template<typename One, bool OneSize = sizeof(One)>
int func() { return OneSize; }

void myfunc()
{
	func<BaseClass>();
}

```

Hmmm, I can see that this is tricky the way the code is designed. Now at least I understand how these things work. It seems we need to store off cached information for these things so we can translate when traversing. I see there is a cache in there as well, once we've reached the "final" type we should also use that cache if it is not already happening.

Evening project for me to get a working prototype it seems :-)

**Comment by honkstar1** [comment]
Ok, I got something that works with templates in templates with typedefs for template types as template type in internal templates... the solution feels, uhm, outside the system. But I guess since it is going on the AST directly the bookkeeping needs to be handled by IWYU.

**Comment by honkstar1** [comment]
Lol, I now went back to the complex scenario we have in Unreal Engine and ofc it didn't work. Here's a simplified version of the scenario

```
#include "BaseClass.h"

template <typename Tpl1, unsigned long Tpl1Size = sizeof(Tpl1)>
struct TGetCastType { static const unsigned long Value = Tpl1Size; };

template <typename Tpl3, unsigned long CastType = TGetCastType<Tpl3>::Value>
struct TCastImpl
{
	static unsigned long GetSize() { return CastType; }
};

template <class Tpl5> unsigned long GetSize() { return TCastImpl<Tpl5>::GetSize(); }

void myfunc()
{
	unsigned long v = GetSize<BaseClass>();(void)v;
}
```

Problem is that when we want to traverse the template argument default initializer we hit a DependentScopeDeclRefExpr since we are traversing non-instantiated path (we must do to be able to find the default initializers).. and I haven't figured out how to find the right expression from that yet. The code is setup to be able to resolve back to real type once I can jump to what that DependentScopeDeclRefExpr refers to.

**Comment by honkstar1** [comment]
Brute-force tracking all NamedDecl to lookup and then traverse that decl from the DependentScopeDeclRefExpr works. Then it will eventually find its way in to the default arguments on the dependent template default initializer.

So that makes above scenario work. This is how clang must do under the hood.. the DependentScopeDeclRefExpr turns in to DeclRefExpr during instantiation so the instantiated part of the ast has DeclRefExpr instead.. but before instantiation it is a DependentScopeDeclRefExpr. I don't know if it is possible to know which NamedDecl that eventually gets queried, but I will see if I can figure it out.

**Comment by kimgr** [comment]
Interesting! I've seen mentions of TreeTransform to explain how template instantiation works in clang. Sounds like that machinery would handle the node replacement you describe here.

**Comment by honkstar1** [comment]
I kind of realized that this is really a can of worms.

IWYU work on 
1. Non-instantiated templates where we can find explicit types in the expressions
2. Instantiated templates where everything is resolved

Unfortunately there is logic in between that translate the non-instantiated templates into instantiated where it actually need types to figure things out... IWYU doesn't see these. Right now it feels like I'm essentially re-creating that step after instantiation where I re-parse the non-instantiated Ast and apply the translation from the instantiated paths to find which types that was used during instantiation.

I don't know what the performance impact will be. Unreal+Fortnite compiles in 2.5h with iwyu.exe (after a few optimizations that makes a big difference that I haven't pushed back here yet). We'll see how much impact these last changes will make.

**Comment by kimgr** [comment]
Hmm, I wonder if the challenges are similar to those for UnresolvedLookupExpr (search issues for more info), I haven't found a way to s through those either.

**Comment by honkstar1** [comment]
ok, considering new approach.. 

This is how it works today
1. We traverse through the tree using IwuAstConsumer
2. When we hit a templated type we do ScanInstantiatedX
3. ScanInstantiatedX run AstFlattenerVisitor
4. AstFlattenerVisitor traverses the normal non-instantiated tree and store all nodes it hit
5. ScanInstantiatedX enter the part of the Ast that is instantiated and skip the ones AstFlattenerVisitor found (since they are parsed by IwuAstConsumer and would just be extra-work revisiting).

My idea is, how about in the AstFlattenerVisitor we check if there is template argument default initializers... store a bool "TreeContainsDefaultInitializers"

If there are default initializers we do:

6. Use the resugar information built in ScanInstantiatedX and re-run another AstFlattenerVisitor called TemplateDefaultTemplateArgumentVisitor where we handle all those expression and get the types.

Right now I'm trying to, as we go through the instantiated template find the way back out to the un-instantiated template and do a little traversal there when I find things with default arguments.. I don't know how much whac-a-mole this solution is.. will continue a little further just to unblock myself. but I think the idea above is cleaner.

Don't know about the TreeTransform stuff. Need to google it and try to understand what it is and if it can be tapped into :-)

**Comment by honkstar1** [comment]
Even better.. this is a simple solution that I think will work

in step 4 we also collect the nodes that are template argument default initializers into a separate list.

and in step 6 we don't have to traverse a list, we can just traverse the ones collected in step 4 using the resugar table.

**Comment by kimgr** [comment]
There's some of that going on already, note how `GetTplType... ` return type => nullptr for default argument types. I'm not 100% sure how that's different from what you propose, but it sounds close enough.

**Comment by honkstar1** [comment]
yes, it doesn't handle nested things. it just gets the outer thing. as soon as you send in template types in template types it falls apart.

**Comment by honkstar1** [comment]
The simple solution with tracking default argument expr from non-instantiated ast and re-traverse after instantiation has been traversed worked!! yay!.. and it feels like a simple, fast (perf wise) solution. It also shouldnt find any additional types (since it just rerun parts of the non-instantiated ast and only resolve the ones it can)

**Comment by kimgr** [comment]
Cool, can't wait to see the patch!

I honestly can't get my ahead around much of the template handling code in IWYU, but were you able to localize this somewhat? From a cohesion perspective it would be nice if all the code related to default template params was close to here: https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_ast_util.cc#L1316

**Comment by honkstar1** [comment]
So I got it working with the full tracking, but it became very expensive.... so I went back to the more local approach and have gotten very far.. I'm confident this will work and be much faster.

There are so many tricky scenarios. consider this

```
template<typename A, bool B = sizeof(A)>
struct  Foo {
   enum { Value = B };
};

void myFunc()
{
   Foo<SomeClass>::B;
}
```

What makes this so tricky is that for the instantiated template path in the Ast Value is just an enum constant. And for the non-instantiated Ast it just goes directly to an EnumDecl... it never goes past any type or anything. The Foo<A, B> is parsed earlier, not in the scaninstantiated stuff... so it is not there at all..

The solution is to use the code that figure out nodes not needed to be parsed.. since that is using the non-instantiated ast path.. and in there we also visit "parents" of vars, enums and funcs to see if they are templated types that has default initializers.

I got all my test code working.. ofc it is not working in the full scenario.. but I will figure it out :-)

**Comment by honkstar1** [comment]
Finally. Last night at 1:30am I managed to get IWYU to output the right includes in my test environments. It was a perfect storm of combinations of default template arguments referencing templated types that had base classes with other default template arguments that in turn had static_asserts with expressions that was sizeof expressions that was templated and returned types of those expression was based on the sent in template types.. Lol, this definitely has been one of the most confusing things I've done in my 25 years as a professional engineer.

Unfortunately now when running through a ton of files in Fortnite/Unreal I get other errors.. so there are more things IWYU doesn't support :-/ This scenario uses variadic templates to implement some sort of std::function similar thing and the generated storage stores a templated array where the dtor is checking if the value has a trivial destructor.

Variadic templates, sigh. let's hope at least that part is implemented in the code :-)

I will dive into this now.

**Comment by kimgr** [comment]
Cool! Yeah, the cross-section of templates and include sanity is surprisingly muddy. Thanks for taking the time!

Let's open a new issue for any problems with variadic templates. There is no explicit handling in IWYU, as far as I know, so probably lots of exciting challenges there.

**Comment by bolshakov-a** [comment]
IWYU analyzes instantiated templates by means of traversing implicit specialization subtrees in the AST. But Clang loses some information when building that subtree, not only _instantiated_ default argument expressions but also C++20 constraints, information about which types were dependent and which were not (that is required to report types not mentioned in template arguments, see bugs #973, #1626). Maybe, Clang maintainers will be kind enough to accept patches adding such information in the AST. Maybe, IWYU should have its own template instantiator (based on or similar to `TreeTransform` in Clang).

As for variadic templates, there were some fixes in IWYU recently.
